File: snippets/go/deduplicate-post-requests-with-idempotency-keys-in-echo/idempotency_middleware.go
package main

import (
	"errors"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"

	"crypto/sha256"
	"encoding/hex"
	"github.com/labstack/echo/v4"
)

type CachedResponse struct {
	StatusCode int
	Headers    http.Header
	Body       []byte
}

type IdempotencyStore interface {
	Load(key string) (CachedResponse, bool)
	TryStart(key string) bool
	Save(key string, response CachedResponse)
	Delete(key string)
}

type memoryEntry struct {
	response   CachedResponse
	inProgress bool
	completed  bool
}

type MemoryIdempotencyStore struct {
	mutex   sync.Mutex
	entries map[string]memoryEntry
}

func NewMemoryIdempotencyStore() *MemoryIdempotencyStore {
	return &MemoryIdempotencyStore{entries: make(map[string]memoryEntry)}
}

func (store *MemoryIdempotencyStore) Load(key string) (CachedResponse, bool) {
	store.mutex.Lock()
	defer store.mutex.Unlock()

	entry, ok := store.entries[key]
	if !ok || !entry.completed {
		return CachedResponse{}, false
	}

	return cloneCachedResponse(entry.response), true
}

func (store *MemoryIdempotencyStore) TryStart(key string) bool {
	store.mutex.Lock()
	defer store.mutex.Unlock()

	entry, exists := store.entries[key]
	if exists && (entry.inProgress || entry.completed) {
		return false
	}

	store.entries[key] = memoryEntry{inProgress: true}
	return true
}

func (store *MemoryIdempotencyStore) Save(key string, response CachedResponse) {
	store.mutex.Lock()
	defer store.mutex.Unlock()

	store.entries[key] = memoryEntry{
		response:  cloneCachedResponse(response),
		completed: true,
	}
}

func (store *MemoryIdempotencyStore) Delete(key string) {
	store.mutex.Lock()
	defer store.mutex.Unlock()

	delete(store.entries, key)
}

func IdempotencyMiddleware(store IdempotencyStore) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			if store == nil {
				return echo.NewHTTPError(http.StatusInternalServerError, "idempotency store is required")
			}
			if c.Request().Method != http.MethodPost {
				return next(c)
			}

			idempotencyKey := strings.TrimSpace(c.Request().Header.Get("Idempotency-Key"))
			if idempotencyKey == "" {
				return echo.NewHTTPError(http.StatusBadRequest, "Idempotency-Key header is required")
			}
			scopedKey := scopedIdempotencyKey(c, idempotencyKey)

			if cachedResponse, ok := store.Load(scopedKey); ok {
				return writeCachedResponse(c.Response().Writer, cachedResponse)
			}
			if !store.TryStart(scopedKey) {
				if cachedResponse, ok := store.Load(scopedKey); ok {
					return writeCachedResponse(c.Response().Writer, cachedResponse)
				}
				return echo.NewHTTPError(http.StatusConflict, "request already in progress")
			}

			// Flow:
			//   require Idempotency-Key on POST
			//      |
			//      +-> completed key -> replay cached response
			//      +-> in-progress key -> return 409
			//      `-> new key -> run handler -> cache successful response -> return it
			response := c.Response()
			originalWriter := response.Writer
			recorder := httptest.NewRecorder()
			response.Writer = recorder

			err := next(c)
			response.Writer = originalWriter
			if err != nil {
				store.Delete(scopedKey)
				return err
			}

			capturedResponse := cachedResponseFromRecorder(recorder)
			if capturedResponse.StatusCode >= http.StatusOK && capturedResponse.StatusCode < http.StatusMultipleChoices {
				store.Save(scopedKey, capturedResponse)
			} else {
				store.Delete(scopedKey)
			}

			if err := writeCachedResponse(originalWriter, capturedResponse); err != nil {
				return fmt.Errorf("write captured response: %w", err)
			}
			return nil
		}
	}
}

func scopedIdempotencyKey(c echo.Context, idempotencyKey string) string {
	requestPath := c.Request().URL.EscapedPath()
	if requestPath == "" {
		requestPath = "/"
	}

	callerScope := requestCallerScope(c)
	return c.Request().Method + ":" + requestPath + ":" + callerScope + ":" + idempotencyKey
}

func requestCallerScope(c echo.Context) string {
	userID := strings.TrimSpace(c.Request().Header.Get("X-User-ID"))
	if userID != "" {
		return "user:" + userID
	}

	authorizationHeader := strings.TrimSpace(c.Request().Header.Get(echo.HeaderAuthorization))
	if authorizationHeader != "" {
		return "auth:" + authorizationScopeHash(authorizationHeader)
	}

	return "anonymous"
}

func authorizationScopeHash(authorizationHeader string) string {
	parts := strings.Fields(authorizationHeader)
	normalizedValue := authorizationHeader
	if len(parts) > 0 {
		normalizedValue = strings.ToLower(parts[0])
		if len(parts) > 1 {
			normalizedValue += " " + strings.Join(parts[1:], " ")
		}
	}

	hash := sha256.Sum256([]byte(normalizedValue))
	return hex.EncodeToString(hash[:16])
}
func cachedResponseFromRecorder(recorder *httptest.ResponseRecorder) CachedResponse {
	statusCode := recorder.Code
	if statusCode == 0 {
		statusCode = http.StatusOK
	}

	headers := make(http.Header, len(recorder.Header()))
	for key, values := range recorder.Header() {
		headers[key] = append([]string(nil), values...)
	}

	body := append([]byte(nil), recorder.Body.Bytes()...)
	return CachedResponse{StatusCode: statusCode, Headers: headers, Body: body}
}

func cloneCachedResponse(response CachedResponse) CachedResponse {
	headers := make(http.Header, len(response.Headers))
	for key, values := range response.Headers {
		headers[key] = append([]string(nil), values...)
	}
	body := append([]byte(nil), response.Body...)
	return CachedResponse{StatusCode: response.StatusCode, Headers: headers, Body: body}
}

func writeCachedResponse(writer http.ResponseWriter, response CachedResponse) error {
	if writer == nil {
		return errors.New("writer is required")
	}
	for key, values := range response.Headers {
		writer.Header()[key] = append([]string(nil), values...)
	}
	writer.WriteHeader(response.StatusCode)
	if len(response.Body) == 0 {
		return nil
	}
	_, err := writer.Write(response.Body)
	return err
}

func newServer(store IdempotencyStore) *echo.Echo {
	e := echo.New()
	e.Use(IdempotencyMiddleware(store))
	e.POST("/orders", func(c echo.Context) error {
		c.Response().Header().Set("X-Order-ID", "order-123")
		return c.JSON(http.StatusCreated, map[string]string{
			"order_id": "order-123",
			"status":   "created",
		})
	})
	e.GET("/orders", func(c echo.Context) error {
		return c.NoContent(http.StatusOK)
	})
	return e
}

func main() {
	store := NewMemoryIdempotencyStore()
	e := newServer(store)
	e.Logger.Fatal(e.Start(":8080"))
}

File: snippets/python/deduplicate-post-requests-with-idempotency-keys-in-fastapi-and-python/src/idempotency_middleware.py
import hashlib
from dataclasses import dataclass
from threading import Lock
from typing import Awaitable, Callable, Dict, Mapping, Optional

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, Response


@dataclass(frozen=True)
class CachedResponse:
    status_code: int
    headers: Dict[str, str]
    body: bytes


@dataclass(frozen=True)
class _MemoryEntry:
    response: Optional[CachedResponse]
    in_progress: bool
    completed: bool


class MemoryIdempotencyStore:
    def __init__(self) -> None:
        self._lock = Lock()
        self._entries: Dict[str, _MemoryEntry] = {}

    def load(self, key: str) -> Optional[CachedResponse]:
        with self._lock:
            entry = self._entries.get(key)
            if entry is None or not entry.completed or entry.response is None:
                return None
            return CachedResponse(
                status_code=entry.response.status_code,
                headers=dict(entry.response.headers),
                body=bytes(entry.response.body),
            )

    def try_start(self, key: str) -> bool:
        with self._lock:
            entry = self._entries.get(key)
            if entry is not None and (entry.in_progress or entry.completed):
                return False
            self._entries[key] = _MemoryEntry(response=None, in_progress=True, completed=False)
            return True

    def save(self, key: str, response: CachedResponse) -> None:
        with self._lock:
            self._entries[key] = _MemoryEntry(
                response=CachedResponse(
                    status_code=response.status_code,
                    headers=dict(response.headers),
                    body=bytes(response.body),
                ),
                in_progress=False,
                completed=True,
            )

    def delete(self, key: str) -> None:
        with self._lock:
            self._entries.pop(key, None)


ProcessOrder = Callable[[], Awaitable[Response]]


def new_app(
    store: Optional[MemoryIdempotencyStore] = None,
    process_order: Optional[ProcessOrder] = None,
) -> FastAPI:
    app = FastAPI()
    active_store = MemoryIdempotencyStore() if store is None else store
    process_order_handler = default_process_order if process_order is None else process_order

    @app.middleware("http")
    async def idempotency_middleware(request: Request, call_next: Callable[[Request], Awaitable[Response]]) -> Response:
        if request.method != "POST":
            return await call_next(request)

        idempotency_key = request.headers.get("Idempotency-Key", "").strip()
        if idempotency_key == "":
            return JSONResponse(status_code=400, content={"detail": "Idempotency-Key header is required"})

        scoped_key = scoped_idempotency_key(request, idempotency_key)
        cached_response = active_store.load(scoped_key)
        if cached_response is not None:
            return response_from_cached(cached_response)
        if not active_store.try_start(scoped_key):
            cached_response = active_store.load(scoped_key)
            if cached_response is not None:
                return response_from_cached(cached_response)
            return JSONResponse(status_code=409, content={"detail": "request already in progress"})

        response = await call_next(request)
        captured_response = await capture_response(response)
        if 200 <= captured_response.status_code < 300:
            active_store.save(scoped_key, captured_response)
        else:
            active_store.delete(scoped_key)

        return response_from_cached(captured_response)

    @app.post("/orders")
    async def create_order() -> Response:
        return await process_order_handler()

    @app.get("/orders")
    async def list_orders() -> Response:
        return Response(status_code=200)

    return app


async def default_process_order() -> Response:
    return JSONResponse(
        status_code=201,
        content={"order_id": "order-123", "status": "created"},
        headers={"X-Order-ID": "order-123"},
    )


def scoped_idempotency_key(request: Request, idempotency_key: str) -> str:
    return ":".join(
        [
            request.method,
            request.url.path or "/",
            request_caller_scope(request.headers),
            idempotency_key,
        ]
    )


def request_caller_scope(headers: Mapping[str, str]) -> str:
    user_id = headers.get("X-User-ID", "").strip()
    if user_id != "":
        return "user:" + user_id

    authorization_header = headers.get("Authorization", "").strip()
    if authorization_header != "":
        return "auth:" + authorization_scope_hash(authorization_header)

    return "anonymous"


def authorization_scope_hash(authorization_header: str) -> str:
    parts = authorization_header.split()
    normalized_value = authorization_header
    if len(parts) > 0:
        normalized_value = parts[0].lower()
        if len(parts) > 1:
            normalized_value += " " + " ".join(parts[1:])
    return hashlib.sha256(normalized_value.encode("utf-8")).hexdigest()[:32]


async def capture_response(response: Response) -> CachedResponse:
    body = b""
    async for chunk in response.body_iterator:
        body += chunk

    return CachedResponse(
        status_code=response.status_code,
        headers={key: value for key, value in response.headers.items()},
        body=body,
    )


def response_from_cached(cached_response: CachedResponse) -> Response:
    headers = {key: value for key, value in cached_response.headers.items() if key.lower() != "content-length"}
    return Response(content=cached_response.body, status_code=cached_response.status_code, headers=headers)


app = new_app()


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)